### PR TITLE
Fix TabSheet attaching and detaching components

### LINKED ISSUE #10985: Grid Buffered Editor bug not fixed with #10146
see my comment in #10146, from 13.06.2018. Since that issue is already closed, I bring up the problem again here.
The fix that originated from #10146 did not fix all issues with editor in grid becoming temporarily invisible, problematic part in TabSheet is even marked with comment //Ugly hack, and this hack has also broken our code in some unrelated places. So please remove the offending #detach();attach() sequence in class TabSheet and consider a different fix.

The following problem still persists: Have a grid, open the editor of one row, then temporarily make the grid invisible by changing the visibility of some uplevel container that contains the grid.

Code to reproduce (copied from original post #10146, with some modifications):
```
Grid grid= new Grid<>(Clientes.class);
grid.removeAllColumns();
grid.getEditor().setEnabled(true);
grid.getEditor().setBuffered(true);
Binder binder = grid.getEditor().getBinder();
Binding<Clientes, String> bind = binder.forField(new TextField()).bind(Clientes::getNombre,Clientes::setNombre);
grid.addColumn(e-> e.getNombre()).setCaption("Nombre").setEditorBinding(bind).setWidth(200);
List lista=new ArrayList<>();
for(int i=0;i<10;i++){
Clientes cli= new Clientes();
cli.setNombre("nombre "+i);
lista.add(cli);
}
grid.setItems(lista);
VerticalLayout tab1 = new VerticalLayout();
tab1.setCaption("tab1");
VerticalLayout tab2 = new VerticalLayout();
tab2.setCaption("tab2");
VerticalLayout tabGrid = new VerticalLayout();
tabGrid.setCaption("GridLayout");
tabGrid.addComponent(grid);
CheckBox cb = new CheckBox("toggle vis");
cb.addValueChangeListener(val -> tabGrid.setVisible(val.getValue()));
tab2.addComponent(cb);
tab2.addComponent(tabGrid);
TabSheet tabSheet= new TabSheet(tab1,tab2);
tabSheet.setSizeFull();
addComponent(tabSheet);
```
The fix in #10146 solves the situation when the grid gets invisible due to switch in tabs. But if you make the grid invisible by toggling the checkbox at top of tab2, with editor open, the editor cannot be opened again when the grid gets visible again.

It would be more consistent if you could close the editor whenever the grid is taken out of client dom, regardless on what has caused this (visibility, Tab switch, whatever...).

Some quick look through invocation sequences has brought two possible places to my attention:

1. 

subclass method `Grid.beforeClientResponse(boolean initial)`, 
and whenever initial==true, close the editor if it is open.

2. 

Another place is ConnectorTracker.clearConnectorMap(), where (around line 336) a component is remarked as removed from browser dom, so you could invent a new hook method named e.g. ClientConnector.removedFromClient(), which does nothing in superclass, but could be overloaded in class Grid to close a pending editor.

This problem was reproduced in vaadin 8.4.2.

### LINKED ISSUE #10987: Ugly hack in #10557 breaks attach() specification of ClientConnector
Since Bugfix #10557 (Vaadin version 8.3.2) Tabsheet calls attach() and detach() when selected tab changes, breaking the attach() [specification](https://vaadin.com/api/com/vaadin/server/ClientConnector.html#attach--) of ClientConnector. 

```
       if (tab != null && !Objects.equals(tab.getComponent(), component) && tab.getComponent() != null && tab.getComponent().isAttached()) {
            tab.getComponent().detach();
            tab.getComponent().attach(); // ugly hack
        }
```
An already attached tab is being detached() an re-attached() again when selected by a user. This breaks the specification of the attach method:

> The caller of this method is Connector#setParent(ClientConnector) if the parent is itself already attached to the session. If not, the parent will call the attach() for all its children when it is attached to the session. This method is always called before the connector's data is sent to the client-side for the first time.

The comments in #10557 already show some users having problems with this unexpected and unspecified behaviour in existing Vaadin applications.

### PR DESCRIPTION
**Pull request by tsuoanttila**
This patch reverts the fix #10557 and replaces it with a
proper solution from Grid perspective.

Fixes #10987
Fixes #10985

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/10988)

### CODE REVIEW

**Comment by elmot** [comment]
Reviewed 6 of 6 files at r1.
Review status: 4 of 6 files reviewed, 1 unresolved discussion (waiting on @elmot and @tsuoanttila)

---

*[server/src/main/java/com/vaadin/ui/Grid.java, line 4789 at r1](https://reviewable.io/reviews/vaadin/framework/10988#-LFHnKmy3C94a2aK1apK:-LFHnKmy3C94a2aK1apL:b4tl4ez) ([raw file](https://github.com/vaadin/framework/blob/b8ed19d5e5597ebb82a4cdd53a941ca4dd06065f/server/src/main/java/com/vaadin/ui/Grid.java#L4789)):*
> ```Java
> 
>     @Override
>     public void detach() {
> ```

Is it correct to move this away from detach? Isn't it safer to cancel editor on every detach rather than make "lazy cancel"?
(Not a blocker)

---

*[uitest/src/test/java/com/vaadin/tests/components/grid/GridInTabSheetTest.java, line 115 at r1](https://reviewable.io/reviews/vaadin/framework/10988#-LFHooZSAdRppd9HUTNk:-LFHooZSAdRppd9HUTNl:bpr2ao8) ([raw file](https://github.com/vaadin/framework/blob/d8bb9895a0cc59c94dee2269290eca69a0b4281f/uitest/src/test/java/com/vaadin/tests/components/grid/GridInTabSheetTest.java#L115)):*
> ```Java
>                 grid.getEditor().getField(1).getAttribute("value"));
>         grid.sendKeys(Keys.ESCAPE);
> ```

Add check if editor is closed here? (Not a blocker)

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10988)*

**Comment by elmot** [comment]
Reviewed 1 of 2 files at r2.
Review status: 5 of 6 files reviewed, 1 unresolved discussion (waiting on @elmot and @tsuoanttila)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10988)*

**Comment by elmot** [comment]
Reviewed 1 of 2 files at r2.
Review status: all files reviewed, 1 unresolved discussion (waiting on @tsuoanttila)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10988)*

**Comment by tsuoanttila** [comment]
Review status: 5 of 7 files reviewed, 1 unresolved discussion (waiting on @elmot and @tsuoanttila)

---

*[server/src/main/java/com/vaadin/ui/Grid.java, line 4789 at r1](https://reviewable.io/reviews/vaadin/framework/10988#-LFHnKmy3C94a2aK1apK:-LFLlqH90I7FsSOY1nKl:b-iw1g9d) ([raw file](https://github.com/vaadin/framework/blob/b8ed19d5e5597ebb82a4cdd53a941ca4dd06065f/server/src/main/java/com/vaadin/ui/Grid.java#L4789)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Is it correct to move this away from detach? Isn't it safer to cancel editor on every detach rather than make "lazy cancel"?
(Not a blocker)

</blockquote></details>

"lazy cancel" is the safe catch all, if it's just detached and never used again, then it's definitely not a problem.

---

*[uitest/src/test/java/com/vaadin/tests/components/grid/GridInTabSheetTest.java, line 115 at r1](https://reviewable.io/reviews/vaadin/framework/10988#-LFHooZSAdRppd9HUTNk:-LFLmNPI1qXBjY_q9tii:boag87p) ([raw file](https://github.com/vaadin/framework/blob/d8bb9895a0cc59c94dee2269290eca69a0b4281f/uitest/src/test/java/com/vaadin/tests/components/grid/GridInTabSheetTest.java#L115)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Add check if editor is closed here? (Not a blocker)
</blockquote></details>

Commented.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10988)*

**Comment by elmot** [comment]
Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10988)*
